From 533ddab94ae3a7a906d9c1391e082c74ceaf4dc2 Mon Sep 17 00:00:00 2001
From: Edouard Cunibil <admin@duael.fr>
Date: Sun, 20 Jul 2014 18:16:53 +0200
Subject: [PATCH] Issue #1946058 by matthensley: Fixed Cannot set field as
 required.

---
 geofield.module |   40 +++++++++++++---------------------------
 1 file changed, 13 insertions(+), 27 deletions(-)

diff --git a/geofield.module b/geofield.module
index d025e2e..19248d2 100644
--- a/geofield.module
+++ b/geofield.module
@@ -158,39 +158,25 @@ function geofield_field_validate($entity_type, $entity, $field, $instance, $lang
   $backend = ctools_get_plugins('geofield', 'geofield_backend', $field['settings']['backend']);
 
   foreach ($items as $delta => $item) {
-    $geom_empty = geofield_geom_is_empty($item);
-
-    // Required field empty.
-    if ($instance['required'] && $geom_empty) {
-      $errors[$field['field_name']][$langcode][$delta][] = array(
-        'error' => 'data_missing', 
-        'message' => t('%name is required and must not be empty.', array('%name' => $instance['label'])),
-      );
-    }
-    else {
-      // Geometry errors.
-      if ($geom_empty) {
-        return FALSE;
+    // Geometry errors.
+    if (!geofield_geom_is_empty($item)) {
+      $error = geofield_validate_geom($item);
+      if ($error) {
+        $errors[$field['field_name']][$langcode][$delta][] = array(
+          'error' => 'data_faulty',
+          'message' => t('%name: Specified location data is invalid.', array('%name' => $instance['label'])),
+        );
       }
-      else {
-        $error = geofield_validate_geom($item);
+
+      if (!empty($backend['validate'])) {
+        $validate_callback = $backend['validate'];
+        $error = $validate_callback($item);
         if ($error) {
           $errors[$field['field_name']][$langcode][$delta][] = array(
-            'error' => 'data_faulty', 
+            'error' => 'data_faulty',
             'message' => t('%name: Specified location data is invalid.', array('%name' => $instance['label'])),
           );
         }
-
-        if (!empty($backend['validate'])) {
-          $validate_callback = $backend['validate'];
-          $error = $validate_callback($item);
-          if ($error) {
-            $errors[$field['field_name']][$langcode][$delta][] = array(
-              'error' => 'data_faulty',
-              'message' => t('%name: Specified location data is invalid.', array('%name' => $instance['label'])),
-            );
-          }
-        }
       }
     }
   }
-- 
1.7.10.4

